iroha_network/src/lib.rs

```diff
@@ -18,37 +19,75 @@ pub type State<T> = Arc<Mutex<T>>;
 pub trait AsyncStream: async_std::io::Read + async_std::io::Write + Send + Unpin {}
 impl<T> AsyncStream for T where T: async_std::io::Read + async_std::io::Write + Send + Unpin {}
 
-pub struct Network {
-    server_url: String,
-}
-
-impl Network {
+#[async_trait]
+pub trait Network {
     /// Creates a new client that will send request to the server on `server_url`
     /// # Arguments
     ///
     /// * `server_url` - is of format ip:port
+    fn new(server_url: &str) -> Self;
+
+    /// Establishes connection to server on `self.server_url`, sends `request` closes connection and returns `Response`.
+    async fn send_request(&self, request: Request) -> Result<Response, String>;
+
+    /// Establishes connection to server on `server_url`, sends `request` closes connection and returns `Response`.
+    async fn send_request_to(server_url: &str, request: Request) -> Result<Response, String>;
+
+    /// Listens on the specified `server_url`.
+    /// When there is an incoming connection, it passes it's `AsyncStream` to `handler`.
+    /// # Arguments
     ///
+    /// * `server_url` - url of format ip:port (e.g. `127.0.0.1:7878`) on which this server will listen for incoming connections.
+    /// * `handler` - callback function which is called when there is an incoming connection, it get's the stream for this connection
+    /// * `state` - the state that you want to capture
+    async fn listen<H, F, S>(
+        state: State<S>,
+        server_url: &str,
+        mut handler: H,
+    ) -> Result<(), String>
+    where
+        H: FnMut(State<S>, Box<dyn AsyncStream>) -> F + Send,
+        F: Future<Output = Result<(), String>> + Send,
+        S: Send;
+
+    /// Helper function to call inside `listen_async` `handler` function to parse and send response.
+    /// The `handler` specified here will need to generate `Response` from `Request`.
+    /// See `listen_async` for the description of the `state`.
+    async fn handle_message_async<H, F, S>(
+        state: State<S>,
+        mut stream: Box<dyn AsyncStream>,
+        mut handler: H,
+    ) -> Result<(), String>
+    where
+        H: FnMut(State<S>, Request) -> F + Send,
+        F: Future<Output = Result<Response, String>> + Send,
+        S: Send;
+}
+
+pub struct IrohaNetwork {
```

Comment:
I also thought about such a names for default impls, but it's not descriptive and I suggest use `TcpNetwork`.